core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1124,19 +1120,20 @@ private void sendResponse(Version nodeVersion, Channel channel, final TransportR
             final TransportResponseOptions finalOptions = options;
             Runnable onRequestSent = () -> { // this might be called in a different thread
                 try {
-                    toRelease.close();
+                    bStream.close();
                 } finally {
                     transportServiceAdapter.onResponseSent(requestId, action, response, finalOptions);
                 }
             };
             addedReleaseListener = internalSendMessage(channel, reference, onRequestSent);
         } finally {
             try {
-                IOUtils.close(stream);
```

Comment:
I think this is too late the close the stream since it might need to flush the content (compressed) to the bytesstream but we send the message async? This entire logic here is super confusing, I don't understand why we can't simply close the top level stream and unless we added a listener given my comment about the non-closing option on the compressor?